sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -328,6 +329,11 @@ class assume_buffer_outlives_graph {
   public:
     assume_buffer_outlives_graph() = default;
 };
+
+class updatable {
+  public:
+    updatable() = default;
+};
```

Comment:
This came up in the implementation PR, but should we have an aspect or device info query for reporting whether a device supports update? Some OpenCL backends could support regular graphs but not update.

Reply:
Is the graph extension very useful without the update ability?  I thought most (all?) of the real use cases required the ability to update nodes in the graph.

Which OpenCL backends can support graphs but not the update feature?

Reply:
> Which OpenCL backends can support graphs but not the update feature?


In OpenCL  [cl_khr_command_buffer](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_command_buffer) is the extension required for basic graph functionality and  [cl_khr_command_buffer_mutable_dispatch](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_command_buffer_mutable_dispatch) is the extension required for update. So it's definitely possible for one vendor to implement one and not the other. PoCL for example, which supports `cl_khr_command_buffer`, but does not implement `cl_khr_command_buffer_mutable_dispatch` (according to http://portablecl.org/docs/html/extensions.html).  [oneAPI Construction Kit](https://github.com/codeplaysoftware/oneapi-construction-kit) supports both but doesn't not implement the optional capabilities in `cl_khr_command_buffer_mutable_dispatch` to update ND-Range, and so therefore can't completely support graph update.

Once we add the ability to change the ability to update the src/des/size of memory copy commands, OpenCL's equivalent API is still being drafted(https://github.com/KhronosGroup/OpenCL-Docs/issues/1066) so no `cl_khr_command_buffer` implementations will be able to support that right now.

> Is the graph extension very useful without the update ability? I thought most (all?) of the real use cases required the ability to update nodes in the graph.

Update is definitely an important feature, but I don't think it's fair to say the extension isn't useful without it. For example reduced latency to submit work the device, even only ever submitting your executable graphs once so never needed to update for a next submission, it could still be beneficial in terms of performance to build a graph of commands in a warmup phase and then submit it in a more latency critical part of the application.

In terms of alternatives, we could emulate this functionality when a backend dosen't support update by recreating the backend command-buffers on `update()`. But it would add latency to the `update()` call and wouldn't be implementable in time for the first delivery of this feature.

Reply:
Another option is to just require any implementation of this extension to implement update also.  This would require certain support in OpenCL, but that seems reasonable to me.  People with an OpenCL that is missing that support can simply choose not to implement the graph API extension at all.

Adding an aspect has a downside because it adds one more thing that applications need to check if they want to be portable.

Reply:
I've also realized that the update feature in L0 might not be available immediately across all L0 devices, so an aspect could let us report when the feature is supported on a L0 device. This should hopefully only be a temporary situation however, so might not be worth making the decision over.

Overall, I definitely agree we want to avoid introducing aspects where possible, so if our bar for backends supporting SYCL-Graphs is being able to perform update then that's fine and will omit the aspect.

Reply:
> I've also realized that the update feature in L0 might not be available immediately across all L0 devices

Is this part of the Level Zero API, or is it just a temporary implementation limitation?  If the Level Zero API has a per-device query telling whether graph update is supported, that might change my opinion.

Reply:
I'll double check with the L0 team to better understand the timeline of how temporary the limitation of not supporting all L0 devices straight away is. The L0 update API is a extension however, so the runtime can query the L0 driver with `zeDriverGetExtensionProperties` to verify if the extension exist for that device.

Reply:
Can `zeDriverGetExtensionProperties` be used to query if an extension is supported for a specific device?  I thought it just told you whether the extension is supported at all (not per-device).

Reply:
sorry yes you are right, that's a query at the platform level rather than device level.

I should have took the to double check the spec https://spec.oneapi.io/level-zero/latest/core/EXT_Exp_MutableCommandList.html#mutable-command-list Which indicates that there is a query mechanism but I'm not very familiar with `zeDeviceGetProperties()`.

> Implementation support for mutable commands may be discovered by providing a [ze_mutable_command_list_exp_properties_t](https://spec.oneapi.io/level-zero/latest/core/api.html#ze-mutable-command-list-exp-properties-t) object in the pNext member of [ze_device_properties_t](https://spec.oneapi.io/level-zero/latest/core/api.html#ze-device-properties-t) in a call to [zeDeviceGetProperties](https://spec.oneapi.io/level-zero/latest/core/api.html#zedevicegetproperties).

Reply:
Yes, you are correct.  That API allows Level Zero to support mutable command lists on some devices, but not others.  Therefore, it seems like we do need some sort of aspect at the SYCL level.

I wonder if the aspect should be "are mutable graphs supported" or "are graphs supported".  Does Level Zero always have the support needed for (unmutable) graphs?

Reply:
Level Zero always has support for graphs as command-lists are a core part of the API, so I think this new aspect should be "are mutable graphs supported". 

However, we also currently have an existing device info query for "are (unmutable) graphs supported",  `ext::oneapi::experimental::info::device::graph_support`, to reflect that fact that not all backends have graphs support (NativeCPU, OpenCL without command-buffer extension). I'm not sure if you'd prefer that was an aspect instead of a device info query to be consistent with a new aspect?